python/lsst/afw/geom/coordinates/coordinates.cc

```diff
@@ -0,0 +1,509 @@
+/*
+ * LSST Data Management System
+ * Copyright 2008-2017  LSST/AURA.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "pybind11/pybind11.h"
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/Extent.h"
+#include "lsst/afw/geom/CoordinateBase.h"
+#include "lsst/afw/geom/CoordinateExpr.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace afw {
+namespace geom {
+namespace {
+
+template <typename Derived, typename T, int N>
+using PyCoordinateBase = py::class_<CoordinateBase<Derived,T,N>>;
+
+template <int N>
+using PyCoordinateExpr = py::class_<CoordinateExpr<N>, CoordinateBase<CoordinateExpr<N>,bool,N>>;
+
+template <typename T, int N>
+using PyExtentBase = py::class_<ExtentBase<T,N>, CoordinateBase<Extent<T,N>,T,N>>;
+
+template <typename T, int N>
+using PyExtent = py::class_<Extent<T,N>, ExtentBase<T,N>>;
+
+template <typename T, int N>
+using PyPointBase = py::class_<PointBase<T,N>, CoordinateBase<Point<T,N>,T,N>>;
+
+template <typename T, int N>
+using PyPoint = py::class_<Point<T,N>, PointBase<T,N>>;
+
+
+template <typename Derived, typename T, int N>
+void declareCoordinateBase(py::module & mod, std::string const & suffix) {
+    static std::string const name = "CoordinateBase" + suffix;
+    PyCoordinateBase<Derived,T,N> cls(mod, name.c_str());
+
+    /* Operators */
+    cls.def("__getitem__", [](CoordinateBase<Derived,T,N> &c, int i) -> T {
+        if (i < 0 || i >= N) {
+            throw py::index_error{};
```

Comment:
Should we be using this instead of the `py::error_already_set` convention you used at https://github.com/lsst/afw/pull/187/files#diff-e11406535f01337997f15de456b1ae3bR176?